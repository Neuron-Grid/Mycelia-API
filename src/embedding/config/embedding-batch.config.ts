export interface EmbeddingBatchConfig {
    defaultBatchSize: number;
    maxBatchSize: number;
    openaiRateLimit: {
        batchSize: number;
        delayMs: number;
        maxRetries: number;
    };
    queue: {
        removeOnComplete: number;
        removeOnFail: number;
        attempts: number;
        backoffDelay: number;
    };
}

export const EMBEDDING_BATCH_CONFIG: EmbeddingBatchConfig = {
    defaultBatchSize: 50,
    maxBatchSize: 100,
    openaiRateLimit: {
        batchSize: 20, // OpenAI embedding API limit
        delayMs: 1000, // 1 second delay between batches
        maxRetries: 3,
    },
    queue: {
        removeOnComplete: 10,
        removeOnFail: 50,
        attempts: 3,
        backoffDelay: 30000, // 30 seconds
    },
} as const;
