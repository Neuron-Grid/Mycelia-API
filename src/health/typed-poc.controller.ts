import { TypedRoute } from "@nestia/core";
import { Controller } from "@nestjs/common";

@Controller("typed-poc")
export class TypedPocController {
    @TypedRoute.Get("ping")
    ping(): { message: string } {
        return { message: "pong" };
    }
}
