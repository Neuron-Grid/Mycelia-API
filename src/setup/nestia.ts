// NOTE: 開発中に Typia トランスフォームが未適用でも起動できるようにする暫定設定
// 本番運用では Typia 変換を有効にすることを推奨
import { NoTransformConfigurationError } from "@nestia/core/lib/decorators/NoTransformConfigurationError";

// 開発・テストのみで無効化（本番は型検証を維持）
if ((process.env.NODE_ENV || "development") !== "production") {
    // 例外スローを無効化（未設定の場合でも JSON.stringify にフォールバック）
    NoTransformConfigurationError.throws = false;
}
