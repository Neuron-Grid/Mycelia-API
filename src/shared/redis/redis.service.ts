import { Inject, Injectable, Logger, OnModuleDestroy } from "@nestjs/common";
import Redis, { RedisOptions } from "ioredis";

type ConnOpts = {
    host: string;
    port: number;
    password?: string;
    db?: number;
    tls?: RedisOptions["tls"];
};

@Injectable()
export class RedisService implements OnModuleDestroy {
    private readonly logger = new Logger(RedisService.name);
    private healthClient?: Redis;

    constructor(
        @Inject("REDIS_CONNECTION_OPTIONS")
        private readonly opts: ConnOpts,
    ) {}

    // 共通オプションを1箇所で合成
    private base(): RedisOptions {
        return {
            host: this.opts.host,
            port: this.opts.port,
            password: this.opts.password,
            db: this.opts.db,
            tls: this.opts.tls,
            // BullMQ推奨設定を追加
            maxRetriesPerRequest: null,
            enableReadyCheck: false,
        };
    }

    // HealthControllerなどが使うメイン接続
    createMainClient(): Redis {
        return new Redis(this.base());
    }

    getHealthClient(): Redis {
        if (!this.healthClient) {
            this.healthClient = this.createHealthClient();
        }
        return this.healthClient;
    }

    private createHealthClient(): Redis {
        const client = new Redis(this.base());
        client.on("error", (error: Error) => {
            this.logger.warn(`Redis health client error: ${error.message}`);
        });
        return client;
    }

    // Bull用クライアント
    // type毎に細かな違いを吸収
    createBullClient(
        type: "client" | "subscriber" | "bclient" = "client",
    ): Redis {
        switch (type) {
            case "client":
                return new Redis(this.base());

            default:
                return new Redis({
                    ...this.base(),
                    enableReadyCheck: false,
                    maxRetriesPerRequest: null,
                });
        }
    }

    async onModuleDestroy(): Promise<void> {
        const clients: Array<Redis | undefined> = [this.healthClient];
        for (const client of clients) {
            if (!client) continue;
            try {
                await client.quit();
            } catch (error) {
                const message =
                    error instanceof Error ? error.message : String(error);
                this.logger.warn(
                    `Failed to close Redis client gracefully: ${message}`,
                );
            }
        }
    }
}
