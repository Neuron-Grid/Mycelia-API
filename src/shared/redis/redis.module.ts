import { Global, Module } from "@nestjs/common";
import { ConfigModule, ConfigService } from "@nestjs/config";
import { RedisService } from "./redis.service";

// どのModuleからも使えるようにGlobalにする
@Global()
@Module({
    imports: [ConfigModule],
    providers: [
        {
            provide: "REDIS_CONNECTION_OPTIONS",
            useFactory: (config: ConfigService) => {
                // 必須: REDIS_URL のみを使用（Heroku固有検出は廃止）
                const url = config.get<string>("REDIS_URL");
                if (!url) throw new Error("REDIS_URL is required");

                const u = new URL(url);

                // TLS は rediss:// の場合のみ有効化（自己署名の特別対応は行わない）
                const tls = u.protocol === "rediss:" ? {} : undefined;

                return {
                    host: u.hostname,
                    port: Number(u.port) || 6379,
                    password: u.password || undefined,
                    db: u.pathname ? Number(u.pathname.slice(1) || 0) : 0,
                    tls,
                };
            },
            inject: [ConfigService],
        },
        RedisService,
    ],
    exports: ["REDIS_CONNECTION_OPTIONS", RedisService],
})
export class RedisModule {}
