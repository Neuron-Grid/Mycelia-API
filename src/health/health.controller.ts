// @file システムヘルスチェックAPIのコントローラ

import { TypedRoute } from "@nestia/core";
import { Controller, UseGuards } from "@nestjs/common";
import { ThrottlerGuard } from "@nestjs/throttler";
import type Redis from "ioredis";
import { AdminRoleGuard } from "@/auth/admin-role.guard";
import { RequiresMfaGuard } from "@/auth/requires-mfa.guard";
import { SupabaseAuthGuard } from "@/auth/supabase-auth.guard";
import type { SuccessResponse } from "@/common/utils/response.util";
import { buildResponse } from "@/common/utils/response.util";
import { RedisService } from "@/shared/redis/redis.service";
import { SupabaseRequestService } from "@/supabase-request.service";
import type { HealthCheckResponseDto } from "./dto/health-check-response.dto";

@Controller("health")
@UseGuards(SupabaseAuthGuard, RequiresMfaGuard, AdminRoleGuard, ThrottlerGuard)
// @public
// @since 1.0.0
export class HealthController {
    // @type {number}
    // @readonly
    // @private
    // @default 5000
    private readonly TIMEOUT_MS = 5000;

    // @param {SupabaseRequestService} supabaseRequestService - Supabaseリクエストサービス
    // @param {Queue} feedQueue - BullMQキュー
    // @param {RedisService} redisService - Redisサービス
    // @since 1.0.0
    // @public
    constructor(
        private readonly supabaseRequestService: SupabaseRequestService,
        private readonly redisService: RedisService,
    ) {}

    // @async
    // @public
    // @since 1.0.0
    // @returns {Promise<HealthCheckResponseDto>} - ヘルスチェック結果
    // @throws {HttpException} - チェック失敗時
    // @example
    // await healthController.checkHealth()
    // @see HealthCheckResponseDto
    @TypedRoute.Get("")
    /** Returns health check results */
    async checkHealth(): Promise<SuccessResponse<HealthCheckResponseDto>> {
        await this.checkDatabaseWithTimeout();
        await this.checkRedisWithTimeout();
        return buildResponse("Health checked", {
            status: "OK",
            db: "OK",
            redis: "OK",
        });
    }
    // @async
    // @private
    // @since 1.0.0
    // @returns {Promise<void>} - DBチェック結果
    // @throws {Error} - DBチェック失敗時
    // @example
    // await healthController['checkDatabaseWithTimeout']()
    // @see SupabaseRequestService.getClient
    private async checkDatabaseWithTimeout(): Promise<void> {
        const supabase = this.supabaseRequestService.getClient();

        await this.withTimeout(
            (async () => {
                const { error } = await supabase
                    .from("users")
                    .select("id")
                    .limit(1);
                if (error) {
                    throw new Error(`Database check failed: ${error.message}`);
                }
            })(),
            this.TIMEOUT_MS,
        );
    }

    // @async
    // @private
    // @since 1.0.0
    // @returns {Promise<void>} - Redisチェック結果
    // @throws {Error} - Redisチェック失敗時
    // @example
    // await healthController['checkRedisWithTimeout']()
    // @see RedisService.createMainClient
    private async checkRedisWithTimeout(): Promise<void> {
        await this.withTimeout(
            (async () => {
                const redisClient = this.redisService.getHealthClient();
                if (redisClient.status === "end") {
                    throw new Error(
                        "Redis health client has been closed unexpectedly",
                    );
                }

                if (this.shouldAwaitReady(redisClient.status)) {
                    await this.waitForReady(redisClient);
                }

                if (redisClient.status !== "ready") {
                    throw new Error(
                        `Redis health client not ready (status: ${redisClient.status})`,
                    );
                }

                const pong = await redisClient.ping();
                if (pong !== "PONG") {
                    throw new Error(
                        `Unexpected Redis ping response: ${pong ?? "<empty>"}`,
                    );
                }
            })(),
            this.TIMEOUT_MS,
        );
    }

    // @private
    // @since 1.0.0
    private shouldAwaitReady(status: Redis["status"]): boolean {
        return [
            "wait",
            "connecting",
            "connect",
            "reconnecting",
            "close",
        ].includes(status);
    }

    // @async
    // @private
    // @since 1.0.0
    private async waitForReady(client: Redis): Promise<void> {
        if (client.status === "ready") {
            return;
        }
        if (client.status === "end") {
            throw new Error("Redis connection ended before becoming ready");
        }

        await new Promise<void>((resolve, reject) => {
            const handleReady = () => {
                cleanup();
                resolve();
            };
            const handleError = (error: Error) => {
                cleanup();
                reject(error);
            };
            const handleEnd = () => {
                cleanup();
                reject(
                    new Error("Redis connection ended before becoming ready"),
                );
            };

            const cleanup = () => {
                client.removeListener("ready", handleReady);
                client.removeListener("error", handleError);
                client.removeListener("end", handleEnd);
            };

            client.once("ready", handleReady);
            client.once("error", handleError);
            client.once("end", handleEnd);

            if (client.status === "ready") {
                handleReady();
            }
        });
    }

    // @async
    // @private
    // @since 1.0.0
    // @template T
    // @param {Promise<T>} promise - タイムアウトをかけるPromise
    // @param {number} ms - タイムアウト時間（ミリ秒）
    // @returns {Promise<T>} - 結果
    // @throws {Error} - タイムアウト時
    // @example
    // await healthController['withTimeout'](Promise.resolve(1), 1000)
    private withTimeout<T>(promise: Promise<T>, ms: number): Promise<T> {
        return new Promise<T>((resolve, reject) => {
            const timer = setTimeout(
                () => reject(new Error(`Timeout after ${ms}ms`)),
                ms,
            );

            promise
                .then((res) => {
                    clearTimeout(timer);
                    resolve(res);
                })
                .catch((err) => {
                    clearTimeout(timer);
                    reject(err);
                });
        });
    }
}
