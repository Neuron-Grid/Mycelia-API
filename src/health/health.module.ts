// @file システムヘルスチェックAPIのNestJSモジュール
// @module
// @public
// @since 1.0.0
// @see ./health.controller

import { Module } from "@nestjs/common";
import { AdminRoleGuard } from "@/auth/admin-role.guard";
import { RequiresMfaGuard } from "@/auth/requires-mfa.guard";
import { SupabaseAuthGuard } from "@/auth/supabase-auth.guard";
import { RedisModule } from "@/shared/redis/redis.module";
import { SupabaseAdminService } from "@/shared/supabase-admin.service";
import { SupabaseRequestModule } from "@/supabase-request.module";
import { HealthController } from "./health.controller";

@Module({
    imports: [SupabaseRequestModule, RedisModule],
    controllers: [HealthController],
    providers: [
        SupabaseAdminService,
        SupabaseAuthGuard,
        RequiresMfaGuard,
        AdminRoleGuard,
    ],
})
export class HealthModule {}
