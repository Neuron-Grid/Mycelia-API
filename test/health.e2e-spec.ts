import { jest } from "@jest/globals";
import "@/setup/nestia";
import {
    type ExecutionContext,
    ForbiddenException,
    INestApplication,
    UnauthorizedException,
} from "@nestjs/common";
import { ConfigService } from "@nestjs/config";
import { Test, TestingModule } from "@nestjs/testing";
import type Redis from "ioredis";
import { AdminRoleGuard } from "@/auth/admin-role.guard";
import { SupabaseAuthGuard } from "@/auth/supabase-auth.guard";
import { HealthController } from "@/health/health.controller";
import { HealthModule } from "@/health/health.module";
import { RedisService } from "@/shared/redis/redis.service";
import { SupabaseAdminService } from "@/shared/supabase-admin.service";
import { SupabaseRequestService } from "@/supabase-request.service";

type AuthMode = "allow" | "unauthorized";

const guardState: {
    authMode: AuthMode;
    mfaSatisfied: boolean;
    adminAllowed: boolean;
} = {
    authMode: "allow",
    mfaSatisfied: true,
    adminAllowed: true,
};

jest.mock("@nestjs/throttler", () => ({
    ThrottlerGuard: class {
        canActivate(): boolean {
            return true;
        }
    },
    Throttle: () => () => undefined,
}));

jest.mock("@/auth/supabase-auth.guard", () => ({
    SupabaseAuthGuard: class {
        canActivate(context: ExecutionContext): boolean {
            if (guardState.authMode === "unauthorized") {
                throw new UnauthorizedException();
            }
            const req = context.switchToHttp().getRequest();
            req.user = { id: "user-1" };
            req.authClaims = {
                role: guardState.adminAllowed ? "admin" : "user",
            };
            return true;
        }
    },
}));

jest.mock("@/auth/requires-mfa.guard", () => ({
    RequiresMfaGuard: class {
        canActivate(): boolean {
            if (!guardState.mfaSatisfied) {
                throw new ForbiddenException(
                    "Multi-factor authentication required",
                );
            }
            return true;
        }
    },
}));

jest.mock("@/auth/admin-role.guard", () => ({
    AdminRoleGuard: class {
        canActivate(): boolean {
            if (!guardState.adminAllowed) {
                throw new ForbiddenException("Admin role required");
            }
            return true;
        }
    },
}));

const createSupabaseStub = () => {
    const limit = jest.fn().mockResolvedValue({ data: [], error: null });
    const select = jest.fn().mockReturnValue({ limit });
    const from = jest.fn().mockReturnValue({ select });
    return { from };
};

async function createApp(): Promise<INestApplication> {
    const supabaseClient = createSupabaseStub();
    const redisClient = {
        status: "ready" as Redis["status"],
        ping: jest.fn().mockResolvedValue("PONG"),
        once: jest.fn().mockReturnThis(),
        removeListener: jest.fn().mockReturnThis(),
    } satisfies Pick<Redis, "status" | "ping" | "once" | "removeListener">;

    const moduleFixture: TestingModule = await Test.createTestingModule({
        imports: [HealthModule],
    })
        .overrideProvider(ConfigService)
        .useValue({
            get: (key: string) => {
                switch (key) {
                    case "REDIS_URL":
                        return "redis://localhost:6379";
                    case "SUPABASE_URL":
                        return "https://example.supabase.co";
                    case "SUPABASE_SERVICE_ROLE_KEY":
                        return "service_role";
                    case "SUPABASE_ANON_KEY":
                        return "anon_key";
                    default:
                        return undefined;
                }
            },
        })
        .overrideProvider(SupabaseRequestService)
        .useValue({ getClient: () => supabaseClient })
        .overrideProvider(SupabaseAdminService)
        .useValue({ getClient: () => ({}) })
        .overrideProvider(RedisService)
        .useValue({ getHealthClient: () => redisClient })
        .compile();

    const app = moduleFixture.createNestApplication();
    await app.init();
    return app;
}

const createExecutionContext = () => {
    const request: Record<string, unknown> = {};
    const context: ExecutionContext = {
        switchToHttp: () => ({
            getRequest: () => request,
        }),
    } as ExecutionContext;
    return { context, request };
};

describe("/api/v1/health guards", () => {
    let app: INestApplication;

    beforeAll(async () => {
        app = await createApp();
    });

    afterAll(async () => {
        await app?.close();
    });

    afterEach(() => {
        guardState.authMode = "allow";
        guardState.mfaSatisfied = true;
        guardState.adminAllowed = true;
    });

    it("throws UnauthorizedException when SupabaseAuthGuard rejects", () => {
        guardState.authMode = "unauthorized";
        const guard = new SupabaseAuthGuard();
        const { context } = createExecutionContext();

        expect(() => guard.canActivate(context)).toThrow(UnauthorizedException);
    });

    it("throws ForbiddenException when AdminRoleGuard rejects", () => {
        guardState.adminAllowed = false;
        const guard = new AdminRoleGuard();
        const { context } = createExecutionContext();

        expect(() => guard.canActivate(context)).toThrow(ForbiddenException);
    });

    it("returns sanitized payload without BullMQ data", async () => {
        const guard = new SupabaseAuthGuard();
        const { context, request } = createExecutionContext();
        guard.canActivate(context);

        const controller = app.get(HealthController);
        const response = await controller.checkHealth();

        expect(response).toMatchSnapshot();
        expect(request.authClaims).toEqual({ role: "admin" });
    });
});
