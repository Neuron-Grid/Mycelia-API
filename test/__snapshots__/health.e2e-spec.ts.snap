// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`/api/v1/health guards returns sanitized payload without BullMQ data 1`] = `
{
  "data": {
    "db": "OK",
    "redis": "OK",
    "status": "OK",
  },
  "message": "Health checked",
}
`;
